### PR TITLE
Improve software CameraInit

### PR DESCRIPTION
**Pull request by gregoire-dl**
## Features list

- [x] Fix a bug in sensor database parsing
- [x] Use `FocalLengthIn35mmFilm` metadata in order to estimate sensor size in cameraInit

### CODE REVIEW

File: src/software/pipeline/main_cameraInit.cpp [hunk 1]
```diff
@@ -396,37 +399,78 @@ int main(int argc, char **argv)
...
-        unknownSensors.emplace(std::make_pair(make, model), view.getImagePath()); // will throw an error message
-        if(!allowIncompleteOutput)
-          continue;
+        const float focalIn35mm = std::stof(view.getMetadata("Exif:FocalLengthIn35mmFilm"));
+        const float imageRatio = std::max(view.getWidth(), view.getHeight()) / std::min(view.getWidth(), view.getHeight());
+
+        if(focalLength > 0)
+        {
+          const float sensorDiag = (focalLength * 43.3) / focalIn35mm; // 43.3 is the diagonal of 35mm film
+          sensorWidth = std::sqrt((1/(1 + imageRatio * imageRatio)) * sensorDiag * sensorDiag);
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: Use floating point constants `1.0/(1.0 + xxxxx) xxxx`

File: src/software/pipeline/main_cameraInit.cpp [hunk 2]
```diff
@@ -396,37 +399,78 @@ int main(int argc, char **argv)
...
+
+
+          ALICEVISION_LOG_INFO("Sensor width computed from 'FocalLength' and 'FocalLengthIn35mmFilm' metadata." << std::endl
+                               << "\t- sensor width: " << sensorWidth << " mm" << std::endl
+                               << "\t- focal length: " << focalLength << " mm");
+        }
+        else
+        {
+          sensorWidth = std::sqrt((1/(1 + imageRatio * imageRatio)) * 43.3 * 43.3);
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: 1.0

File: src/software/pipeline/main_cameraInit.cpp [hunk 3]
```diff
@@ -396,37 +399,78 @@ int main(int argc, char **argv)
...
+          focalLength = sensorWidth * (focalIn35mm ) / 36;
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: 36.0

Comment (simogasp) [reply src/software/pipeline/main_cameraInit.cpp]: `36.0f`

File: src/aliceVision/sensorDB/Datasheet.cpp [hunk 4]
```diff
@@ -7,70 +7,43 @@
...
-  std::string brandlower = _brand;
-  boost::algorithm::to_lower(brandlower);
+  std::string brandA = _brand;
+  std::string brandB = other._brand;
 
-  for(const auto& brand : vec_brand)
-  {
-    std::string brandlower2 = brand;
-    boost::algorithm::to_lower(brandlower2);
+  std::transform(brandA.begin(), brandA.end(), brandA.begin(), ::tolower); //tolower
```
Comment (fabiencastan) [inline src/aliceVision/sensorDB/Datasheet.cpp]: As done before, the boost version is more straightforward: `boost::algorithm::to_lower(brandA);`

File: src/aliceVision/sensorDB/parseDatabase.cpp [hunk 5]
```diff
@@ -56,16 +56,14 @@ bool parseDatabase(const std::string& databaseFilePath, std::vector<Datasheet>&
 
 bool getInfo(const std::string& brand, const std::string& model, const std::vector<Datasheet>& databaseStructure, Datasheet& datasheetContent)
 {
-  bool existInDatabase = false;
-
   Datasheet refDatasheet(brand, model, -1.);
   auto datasheet = std::find(databaseStructure.begin(), databaseStructure.end(), refDatasheet);
   if(datasheet != databaseStructure.end())
```
Comment (fabiencastan) [inline src/aliceVision/sensorDB/parseDatabase.cpp]:
```
if(datasheet == databaseStructure.end())
    return false;
```

File: src/aliceVision/sfm/viewIO.hpp [hunk 6]
```diff
@@ -33,6 +34,7 @@ void updateIncompleteView(View& view);
  * @return shared_ptr IntrinsicBase
  */
 std::shared_ptr<camera::IntrinsicBase> getViewIntrinsic(const View& view,
+                                                float mmFocalLength = -1.0f,
```
Comment (fabiencastan) [inline src/aliceVision/sfm/viewIO.hpp]: double

File: src/aliceVision/sfm/viewIO.cpp [hunk 7]
```diff
@@ -64,6 +64,7 @@ void updateIncompleteView(View& view)
 }
 
 std::shared_ptr<camera::IntrinsicBase> getViewIntrinsic(const View& view,
+                                                        float mmFocalLength,
```
Comment (fabiencastan) [inline src/aliceVision/sfm/viewIO.cpp]: double

File: src/software/pipeline/main_cameraInit.cpp [hunk 8]
```diff
@@ -362,9 +363,11 @@ int main(int argc, char **argv)
     View& view = *(std::next(viewPairItBegin,i)->second);
     IndexT intrinsicId = view.getIntrinsicId();
     double sensorWidth = -1;
+    float focalLength = view.getMetadataFocalLength();
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: double

File: src/software/pipeline/main_cameraInit.cpp [hunk 9]
```diff
@@ -396,37 +399,78 @@ int main(int argc, char **argv)
...
-        #pragma omp critical
-        unknownSensors.emplace(std::make_pair(make, model), view.getImagePath()); // will throw an error message
-        if(!allowIncompleteOutput)
-          continue;
+        const float focalIn35mm = std::stof(view.getMetadata("Exif:FocalLengthIn35mmFilm"));
+        const float imageRatio = std::max(view.getWidth(), view.getHeight()) / std::min(view.getWidth(), view.getHeight());
+
+        if(focalLength > 0.0f)
+        {
+          const float sensorDiag = (focalLength * 43.3) / focalIn35mm; // 43.3 is the diagonal of 35mm film
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: double

File: src/software/pipeline/main_cameraInit.cpp [hunk 10]
```diff
@@ -396,37 +399,78 @@ int main(int argc, char **argv)
...
+
+
+          ALICEVISION_LOG_INFO("Sensor width computed from 'FocalLength' and 'FocalLengthIn35mmFilm' metadata." << std::endl
+                               << "\t- sensor width: " << sensorWidth << " mm" << std::endl
+                               << "\t- focal length: " << focalLength << " mm");
+        }
+        else
+        {
+          sensorWidth = std::sqrt((1.0/(1.0 + imageRatio * imageRatio)) * 43.3 * 43.3);
+          focalLength = sensorWidth * (focalIn35mm ) / 36.0f;
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: double and 36.0

File: src/software/pipeline/main_cameraInit.cpp [hunk 11]
```diff
@@ -476,6 +520,20 @@ int main(int argc, char **argv)
       ALICEVISION_LOG_WARNING("\t- '" << imagePath << "'");
   }
 
+
+  if(!unsureSensors.empty())
+  {
+    ALICEVISION_LOG_WARNING("The camera found in the database is slightly different for image(s):");
+    for(const auto& unsureSensor : unsureSensors)
+      ALICEVISION_LOG_WARNING("image: '" << fs::path(unsureSensor.second.first).filename().string() << "'" << std::endl
+                        << "\t- camera brand: " << unsureSensor.first.first <<  std::endl
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: image camera brand
image camera model

File: src/software/pipeline/main_cameraInit.cpp [hunk 12]
```diff
@@ -396,37 +399,78 @@ int main(int argc, char **argv)
...
-      else
+
+      // try to find / compute with 'FocalLengthIn35mmFilm' metadata
+      if(sensorWidth == -1.0 && hasFocalIn35mmMetadata)
       {
-        #pragma omp critical
-        unknownSensors.emplace(std::make_pair(make, model), view.getImagePath()); // will throw an error message
-        if(!allowIncompleteOutput)
-          continue;
+        const float focalIn35mm = std::stof(view.getMetadata("Exif:FocalLengthIn35mmFilm"));
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: double

File: src/software/pipeline/main_cameraInit.cpp [hunk 13]
```diff
@@ -396,37 +399,78 @@ int main(int argc, char **argv)
...
-        unknownSensors.emplace(std::make_pair(make, model), view.getImagePath()); // will throw an error message
-        if(!allowIncompleteOutput)
-          continue;
+        const double focalIn35mm = std::stod(view.getMetadata("Exif:FocalLengthIn35mmFilm"));
+        const double imageRatio = std::min(view.getWidth(), view.getHeight()) / std::max(view.getWidth(), view.getHeight());
+
+        if(focalLength > 0.0)
+        {
+          const double sensorDiag = (focalLength * 43.3) / focalIn35mm; // 43.3 is the diagonal of 35mm film
+          sensorWidth = std::sqrt((1.0/(1.0 + imageRatio * imageRatio)) * sensorDiag * sensorDiag);
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: Remove the useless `1.0/`:
`sensorWidth = std::sqrt((sensorDiag * sensorDiag) / (1.0 + imageRatio * imageRatio));`

File: src/software/pipeline/main_cameraInit.cpp [hunk 14]
```diff
@@ -396,37 +399,78 @@ int main(int argc, char **argv)
...
+      // try to find / compute with 'FocalLengthIn35mmFilm' metadata
+      if(sensorWidth == -1.0 && hasFocalIn35mmMetadata)
       {
-        #pragma omp critical
-        unknownSensors.emplace(std::make_pair(make, model), view.getImagePath()); // will throw an error message
-        if(!allowIncompleteOutput)
-          continue;
+        const double focalIn35mm = std::stod(view.getMetadata("Exif:FocalLengthIn35mmFilm"));
+        const double imageRatio = std::min(view.getWidth(), view.getHeight()) / std::max(view.getWidth(), view.getHeight());
+
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp]: Precompute exact 24x36 diag value:
`const double diag24x36 = std::sqrt(36.0*36.0 + 24.0*24.0);`

File: src/software/pipeline/main_cameraInit.cpp (hunk 13 above)
Comment (simogasp) [reply src/software/pipeline/main_cameraInit.cpp]: or even 
```
sensorWidth = sensorDiag * std::sqrt(1.0 / (1.0 + imageRatio * imageRatio));
```

File: src/aliceVision/sfm/View.hpp [hunk 15]
```diff
@@ -251,12 +251,12 @@ class View
    * @brief Get the corresponding "FocalLength" metadata value
    * @return the metadata value float or -1 if no corresponding value
    */
-  float getMetadataFocalLength() const
+  double getMetadataFocalLength() const
   {
     if(hasMetadata("Exif:FocalLength"))
-      return std::stof(getMetadata("Exif:FocalLength"));
+      return std::stod(getMetadata("Exif:FocalLength"));
```
Comment (fabiencastan) [inline src/aliceVision/sfm/View.hpp]: Can you add a check to return the value only if it's not 0?
If the value is zero we consider that the metadata doesn't exist.

Same for FocalLengthIn35mm.

**Pull request merged**